runtime/health.go

```diff
@@ -57,6 +52,163 @@ func (r *Runtime) Health(ctx context.Context) (*Health, error) {
 	}, nil
 }
 
-func (r *Runtime) InstanceHealth(ctx context.Context, instanceID string) (*InstanceHealth, error) {
-	return r.registryCache.instanceHealth(ctx, instanceID)
+func (r *Runtime) InstanceHealth(ctx context.Context, instanceID string, query DashboardHealthQuery) (*InstanceHealth, error) {
+	res := &InstanceHealth{}
+	// check repo error
+	repo, rr, err := r.Repo(ctx, instanceID)
+	if err != nil {
+		res.Repo = err.Error()
+	} else {
+		err = repo.(drivers.Handle).Ping(ctx)
+		if err != nil {
+			res.Repo = err.Error()
+		}
+		rr()
+	}
+
+	ctrl, err := r.Controller(ctx, instanceID)
+	if err != nil {
+		res.Controller = err.Error()
+		return res, nil
+	}
+
+	// check olap error
+	olap, release, err := r.OLAP(ctx, ctrl.InstanceID, "")
+	if err != nil {
+		res.OLAP = err.Error()
+		return res, nil
+	}
+	defer release()
+
+	resources, err := ctrl.List(ctx, ResourceKindMetricsView, "", false)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheEnabled := healthCheckCacheEnabled(olap)
+	if cacheEnabled {
+		instanceHealth, ok := r.cachedInstanceHealth(ctx, ctrl.InstanceID, ctrl.catalog.version, resources)
+		if ok {
+			res.OLAP = instanceHealth.OLAP
+			res.Dashboards = instanceHealth.Dashboards
+			return res, nil
+		}
+	}
+
+	err = olap.(drivers.Handle).Ping(ctx)
+	if err != nil {
+		res.OLAP = err.Error()
+	} else if query != nil {
+		// run queries against dashboards
+		res.Dashboards = make(map[string]string, len(resources))
+		for _, mv := range resources {
+			q, err := query(ctx, ctrl.InstanceID, mv.Meta.Name.Name)
+			if err != nil {
+				res.Dashboards[mv.Meta.Name.Name] = err.Error()
+				continue
+			}
+			err = r.Query(ctx, ctrl.InstanceID, q, 1)
+			if err != nil {
+				res.Dashboards[mv.Meta.Name.Name] = err.Error()
+			}
+		}
+	}
+	if !cacheEnabled {
+		return res, nil
+	}
+
+	// save to cache
+	hash, err := healthResultHash(ctrl.catalog.version, resources)
+	if err != nil {
+		return nil, err
+	}
+	res.Hash = hash
+
+	bytes, err := json.Marshal(res)
+	if err != nil {
+		return nil, err
+	}
+
+	catalog, release, err := r.Catalog(ctx, instanceID)
+	if err != nil {
+		return nil, err
+	}
+	defer release()
+
+	err = catalog.UpsertInstanceHealth(ctx, &drivers.InstanceHealth{
+		InstanceID: instanceID,
+		Health:     bytes,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (r *Runtime) cachedInstanceHealth(ctx context.Context, instanceID string, ctrlVersion int64, resources []*runtimev1.Resource) (*InstanceHealth, bool) {
+	catalog, release, err := r.Catalog(ctx, instanceID)
+	if err != nil {
+		return nil, false
+	}
+	defer release()
+
+	cached, err := catalog.FindInstanceHealth(ctx, instanceID)
+	if err != nil {
+		return nil, false
+	}
+
+	c := &InstanceHealth{}
+	err = json.Unmarshal(cached.Health, c)
+	if err != nil {
+		return nil, false
+	}
+
+	hash, err := healthResultHash(ctrlVersion, resources)
```

Comment:
This computes the hash using the current controller version, not the one used for the cached health check. 

If the health check itself is refactored to check the cache per-metrics view (instead of for the whole instance) as discussed above, then I think this could be simplified so it doesn't need a hash. Instead, it should store the controller version as part of the cached health check. For example, the `instances` table might have an additional `health_check_controller_version` field or something like that.

Thinking about the high-level goal here, maybe these can be treated as separate steps:
1. Caching the most recent health check
2. Using the cached health check instead of running a new query when a) the controller version hasn't changed, b) the OLAP may be scaled to zero, c) the metrics view is present in the cache.

Reply:
> This computes the hash using the current controller version, not the one used for the cached health check.

I did not understand this. The idea was to discard cache if controller is restarted so that's why controller version is part of the hash along with the state version of all metrics views. Below this line we will compare the new hash with the hash stored in the health check.
Though a controller restart will reconcile all metrics views which will update their state version but it should still be okay to keep a controller version if the underlying metrics view reconciler logic is updated.

For caching at a metrics_view level also the current approach works with following changes:
1. `Hash` is just the controller version.
2. If OLAP is not scaled to zero always trigger health check.
3. Along with the cached `metrics_view` health also store the state.version to check whether cache can be used.

Reply:
> I did not understand this. 

Sorry my bad! I seem to have got the logic wrong here.

> Hash is just the controller version.

Yes – then let's also rename it to `controller_version` for clarity :) 

> Along with the cached metrics_view health also store the state.version to check whether cache can be used.

Makes sense